compiler/optimizer/VPHandlers.cpp

```diff
@@ -4870,6 +4870,9 @@ static void devirtualizeCall(OMR::ValuePropagation *vp, TR::Node *node)
       return;
       }
 
+   if (!vp->comp()->fe()->canDevirtualizeDispatch())
+      return;
+
```

Comment:
Is the situation you described (where a virtual call is not de-virtualized and yet has it's CP index changed to -1) one where we convert an interface call to a virtual call ? 

Regardless, is it avoiding precisely that case that cannot be handled by AOT or also avoiding cases when it could have devirtualized without causing any issues wrt AOT ? e.g. the simplest case when the method mentioned in the virtual call is changed to direct call of the same method, i.e. no new CP index is needed ?

Reply:
>Is the situation you described (where a virtual call is not de-virtualized and yet has it's CP index changed to -1) one where we convert an interface call to a virtual call ?

I don't know about that. I would have to re-try running my tests with this code removed to see what the condition was that resulted in a virtual call where the J2I Thunk had a cpIndex of -1.

>Regardless, is this avoiding precisely that case that cannot be handled by AOT or also avoiding cases when it could have devirtualized without causing any issues wrt AOT ? e.g. the simplest case when the method mentioned in the virtual call is changed to direct call of the same method, i.e. no new CP index is needed ?

It is definitely avoiding the case that cannot be handled by AOT. However, I don't know if this now also avoids the cases when it could have been devirtualized; it is something that I moved on from due to time constraints and so I suppose there is opportunity to not be as conservative as I have been here. That said, this change is consistent with other parts of (at least OpenJ9) code that check do a front end query to see if devirtualization is possible.

I'll open an issue to revisit this.

Reply:
I suspect this is because the J2I virtual thunk relocations work based on a constant pool entry. To properly interact with the "symbol validation manager," they should instead take a method ID.